### ISSUE TITLE
Ability to use system installed package

### ISSUE THREAD

**Issue by gzagaris** [LLNL/spack#1505]
I haven't been able to find a good way to tell spack to use a system installed package. 

Is there a good way to override spack for example to tell it to use the system installed python, MPI or Doxygen?

Best,
George

**Comment by adamjstewart** [comment]
http://software.llnl.gov/spack/configuration.html?highlight=external#external-packages

**Comment by ashkan2200** [comment]
@adamjstewart I tried the packages.yaml but I get errors.
In the ~/.spack/packages.yaml I have added 

    packages:
        openmpi:
            paths:
                 openmpi@2.0.0%gcc@6.1.0 arch=linux-redhat6-x86_64: /opt/openmpi/2.0.0gcc6.1.0 
        cmake:
            paths:
                 cmake@2.8.12.2%gcc@6.1.0 arch=linux-redhat6-x86_64: /usr/bin/cmake


but when I try installing anything, I get this trace

     Traceback (most recent call last):
       File "/home/alid/spack/bin/spack", line 184, in <module>
           main()
       File "/home/alid/spack/bin/spack", line 161, in main
           return_val = command(parser, args)
      File "/pica/h1/alid/spack/lib/spack/spack/cmd/install.py", line 79, in install
           specs = spack.cmd.parse_specs(args.packages, concretize=True)
      File "/pica/h1/alid/spack/lib/spack/spack/cmd/__init__.py", line 106, in parse_specs
           spec.concretize()  # implies normalize
      File "/pica/h1/alid/spack/lib/spack/spack/spec.py", line 1252, in concretize
           self._expand_virtual_packages(),
      File "/pica/h1/alid/spack/lib/spack/spack/spec.py", line 1169, in _expand_virtual_packages
           spec)
     File "/pica/h1/alid/spack/lib/spack/spack/concretize.py", line 116, in     choose_virtual_or_external
          candidates = self._valid_virtuals_and_externals(spec)
     File "/pica/h1/alid/spack/lib/spack/spack/concretize.py", line 108, in  _valid_virtuals_and_externals
          usable.sort(cmp=cmp_externals)
     File "/pica/h1/alid/spack/lib/spack/spack/concretize.py", line 96, in cmp_externals
          return candidates.index(a) - candidates.index(b)
     ValueError: list.index(x): x not in list

**Comment by davydden** [comment]
@ashkan2200 i think you miss `version`, i.e.

```
mkl:
  version: [11.3.3.210]
  paths:
     mkl@11.3.3.210: /opt/intel/mkl
  buildable: False
```
`buildable` is optional for sure.

**Comment by davydden** [comment]
but you are right that it is not mentioned in the documentation. Puzzled...

**Comment by ashkan2200** [comment]
@davydden I still get the same error. If I try adding `module` instead of `paths` I get another error which I'll post in a minute

**Comment by ashkan2200** [comment]
Here is the error, unless I am mistaken again, modules refer to what you would call in `module load ...`

    Traceback (most recent call last):
      File "/home/alid/spack/bin/spack", line 81, in <module>
        import spack
      File "/pica/h1/alid/spack/lib/spack/spack/__init__.py", line 104, in <module>
        pkgsort = PreferredPackages()
      File "/pica/h1/alid/spack/lib/spack/spack/preferred_packages.py", line 35, in __init__
        self.preferred = spack.config.get_config('packages')
      File "/pica/h1/alid/spack/lib/spack/spack/config.py", line 400, in get_config
        data = scope.get_section(section)
      File "/pica/h1/alid/spack/lib/spack/spack/config.py", line 245, in get_section
        data   = _read_config_file(path, schema)
      File "/pica/h1/alid/spack/lib/spack/spack/config.py", line 323, in _read_config_file
        validate_section(data, schema)
      File "/pica/h1/alid/spack/lib/spack/spack/config.py", line 217, in validate_section
        raise ConfigFormatError(e, data)
    spack.config.ConfigFormatError: /home/alid/.spack/packages.yaml:4: 'openmpi/2.0.0' is not of     type 'object'

**Comment by ashkan2200** [comment]
And here is what I have now in my packages.yaml

    packages:
        openmpi:
            version: [2.0.0]
            paths:
                openmpi@2.0.0%gcc@6.1.0 arch=linux-redhat6-x86_64: /opt/openmpi/2.0.0gcc6.1.0
            buildable: False
        cmake:
            version: [2.8.12.2]
            paths:
                cmake@2.8.12.2%gcc@6.1.0 arch=linux-redhat6-x86_64: /usr/bin/cmake
            buildable: False

It somehow doesn't make sense to repeat the version both in `version:[2.0.0]` and in `openmpi@2.0.0`

**Comment by davydden** [comment]
@ashkan2200 i suppose it's because you could have different versions externally so maybe it's just a quick hook to check if a desired version available before going into `paths` section.

**Comment by tomgreen66** [comment]
Just to add - I get a similar error:
```
Traceback (most recent call last):
  File "./bin/spack", line 184, in <module>
    main()
  File "./bin/spack", line 161, in main
    return_val = command(parser, args)
  File "/home/sistg1/work/repos/arcca-git/spack/lib/spack/spack/cmd/spec.py", line 57, in spec
    spec.concretize()
  File "/home/sistg1/work/repos/arcca-git/spack/lib/spack/spack/spec.py", line 1253, in concretize
    self._expand_virtual_packages(),
  File "/home/sistg1/work/repos/arcca-git/spack/lib/spack/spack/spec.py", line 1170, in _expand_virtual_packages
    spec)
  File "/home/sistg1/work/repos/arcca-git/spack/lib/spack/spack/concretize.py", line 116, in choose_virtual_or_external
    candidates = self._valid_virtuals_and_externals(spec)
  File "/home/sistg1/work/repos/arcca-git/spack/lib/spack/spack/concretize.py", line 108, in _valid_virtuals_and_externals
    usable.sort(cmp=cmp_externals)
  File "/home/sistg1/work/repos/arcca-git/spack/lib/spack/spack/concretize.py", line 96, in cmp_externals
    return candidates.index(a) - candidates.index(b)
ValueError: list.index(x): x not in list
```
When I have setup a package intelmpi to point to our already installed Intel MPI installation.  My package.yaml is:
```
packages:
  intelmpi:
    version: [4.1.3]
    paths:
      intelmpi@4.1.3 %gcc@4.4.7 arch=linux-redhat6-x86_64: /software/compilers/intel/2013.1.046/impi/4.1.3.048/intel64
    buildable: False
  all:
    compiler: [gcc@4.4.7, intel@15.0.0]
    providers:
      mpi: [intelmpi]
```
The list of MPI providers is:
```
$ ./bin/spack providers mpi
    intel-parallel-studio@cluster:+mpi  intelmpi@4.1:  mpich@3:      mvapich2@2.0:  openmpi@1.7.5:
    intelmpi@4.0:                       mpich@1:       mvapich2@1.9  openmpi@1.6.5  openmpi@2.0.0:
```
And the IntelMPI is a basic package which provides a number of versions (similar to existing MPI packages).
When debugging the problem it seemed the concretize.py was using the complete name from the paths option (e.g. intelmpi@4.1.3%gcc@4.4.7 arch ...etc) whilst the list of possible MPI providers only has intelmpi@4.0: and intelmpi@4.1: which then doesnt match the paths option version.

**Comment by tomgreen66** [comment]
Also not convinced by (or understand) commit f96c979 - maybe it was trying to skip the code block for any external package, so instead it should be:
```python
if a.name != b.name and (not a.external and not a.external_module and
                         not b.external and not b.external_module):
```

**Comment by ashkan2200** [comment]
So I looked around to see what the problem is,
in the file `spack/lib/spack/spack/concretize.py` in function `_valid_virtuals_and_externals` there is this line `candidates = sorted(providers, cmp=provider_cmp)`

In the line I just mentioned, some of the versions are removed. For instance in my case, in the list of candidates there is no `openmpi@2.0.1` anymore after the sort. 

I removed that line and the exception went away. Maybe this is of some help.
